tests/topology_test.cpp

```diff
@@ -0,0 +1,162 @@
+#include <catch2/catch_test_macros.hpp>
+#include <catch2/matchers/catch_matchers.hpp>
+#include <catch2/matchers/catch_matchers_string.hpp>
+
+#include "vectorcxx/src/lib.rs.h"
+#include <cpr/cpr.h>
+#include <exception>
+#include <filesystem>
+#include <iostream>
+#include <nlohmann/json.hpp>
+#include <spdlog/spdlog.h>
+#include <thread>
+#include <fmt/format.h>
+
+using Catch::Matchers::ContainsSubstring;
+
+namespace {
+  // this is the default file sink used by all testing configs
+  inline auto FILE_SINK_PATH = std::filesystem::path("/tmp/vector_test_sink.log");
+  inline auto DATA_DIR = std::filesystem::path("/tmp/vector");
+
+  // get a file path under data folder
+  std::filesystem::path _file_path(const std::string &file_name) {
+    return std::filesystem::path(__FILE__).parent_path() / "data" / file_name;
+  }
+
+  std::string _load_config(const std::string &file_name) {
+    std::ifstream stream(_file_path(file_name + ".json"));
+    std::stringstream buffer;
+    buffer << stream.rdbuf();
+    auto file_content = buffer.str();
+    // validate the file is a valid json
+    auto _ = nlohmann::json::parse(file_content);
+    return file_content;
+  }
+
+  std::vector<std::string> _read_file_by_lines(const std::filesystem::path &file_path) {
+    std::ifstream input(file_path);
+    std::vector<std::string> lines;
+    for (std::string line; getline(input, line);) {
+      lines.emplace_back(std::move(line));
+    }
+    return lines;
+  }
+
+  std::vector<std::string>
+  _read_events_from_sink(const std::filesystem::path &file_path = FILE_SINK_PATH) {
+    return _read_file_by_lines(file_path);
+  }
+
+  void _wait(uint32_t milliseconds = 200) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(milliseconds));
+  }
+
+  void send_http_events(const std::vector<std::string> &events, uint32_t port = 9999) {
```

Comment:
I switch to use http source for testing since it is simpler than problematic stdin